goss-testing/scripts/data-validator.py

```diff
@@ -0,0 +1,289 @@
+  
+  return filtered_data
+
+
+def drill_down_data(data, desired_keys):
+  for i in range(0, ( len(desired_keys) + 1), 1):
+    if i == len(desired_keys):
+      return saved_data
+    if i == 0:
+      saved_data = data[desired_keys[i]]
+    if i != 0:
+      saved_data = saved_data[desired_keys[i]]
+
+
+def are_valid_ip_masks(data, desired_keys):
+  """Checks that ip/mask or cidr is valid.
+  """
+  filtered_data = user_data(data)
+  err = 0
+
+  for blob in filtered_data:
+    instance_hostname = blob['hostname']
+    child_key = drill_down_data(blob, desired_keys)
+
+    if not child_key:
+      print_err(f"ERR: {desired_keys} is not defined for: {instance_hostname}")
+      err = 1
+    else:
+      for value in child_key:
+        try:
+          ipaddress.IPv4Network(value, strict=False)
+        except ValueError:
+          print_err(f"ERR: {instance_hostname}: '{value}' is not a valid ip/mask in {desired_keys}")
+          err = 1
+  if err == 1:
+    return err
+
+def check_hostname_syntax(hostname):
+  """Checks that a given hostname syntax is valid.
+  """  
+  if len(hostname) > 253:
+      return False
+  if hostname[-1] == ".":
+      hostname = hostname[:-1]
+  allowed = re.compile("(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
+  return all(allowed.match(x) for x in hostname.split("."))
+
+
+def are_valid_hostnames(data, desired_keys):
+  """Checks that a list of hostnames is defined and valid.
+  """
+  filtered_data = user_data(data)
+  err = 0
+
+  for blob in filtered_data:
+    instance_hostname = blob['hostname']
+    child_key = drill_down_data(blob, desired_keys)
+
+    if not child_key:
+      print_err(f"ERR: {desired_keys} is not defined for: {instance_hostname}: ")
+      err = 1
+    else:
+      for item in drill_down_data(blob, desired_keys):
+        if not check_hostname_syntax(item):
+          print_err(f"ERR: {instance_hostname}: '{item}' is not a valid hostname in {desired_keys}")
+          err = 1
+  return err
+
+
+def are_hosts_sane(data, desired_keys):
+  """Checks if a value is sane for a given system.
+     Needs to check more stuff.
+  """
+  filtered_data = user_data(data)
+  target_list = []
+  hosts_list = []
+  err = 0
+
+  if are_valid_hostnames(data, desired_keys):
+    err = 1
+
+  for blob in filtered_data:
+    hosts_list.append(blob['hostname'])
+    for value in drill_down_data(blob, desired_keys):
+      target_list.append(value)
+  # the following may or may not be bad configuration. TBD
+  #    if value == blob['hostname']:
+  #      print_err(f"WARN: {blob['hostname']}: should not reference itself in {desired_keys}")
+  #      err = 1
+
+  # Check that a host definition exists for nodes in peers/servers
+  # needs to be fixed to not show items like ntp.hpecorp.net
+  for i in set(target_list).difference(set(hosts_list)):
+    if re.search("ncn-", i):
+      print_err(f"WARN: {i}: defined in {desired_keys}, but host not defined in BSS / Basecamp" % (i, desired_keys))
```

Comment:
You don't need that % if using an fstring here, I don't think.

```suggestion
      print_err(f"WARN: {i}: defined in {desired_keys}, but host not defined in BSS / Basecamp")
```